starfish/codebook.py

```diff
@@ -513,24 +532,33 @@ def _view_row_as_element(array: np.ndarray) -> np.ndarray:
                      'formats': ncols * [array.dtype]}
             return array.view(dtype)
 
+        self._validate_decode_intensity_input_matches_codebook_shape(intensities)
+
         max_channels = intensities.argmax(Indices.CH.value)
         codes = self.argmax(Indices.CH.value)
 
+        # calculate distance scores by evaluating the fraction of signal in each round that is
+        # found in the non-maximal channels.
+        max_intensities = intensities.max(Indices.CH.value)
+        round_intensities = intensities.sum(Indices.CH.value)
+        distance = 1 - (max_intensities / round_intensities).mean(Indices.ROUND.value)
+
         a = _view_row_as_element(codes.values.reshape(self.shape[0], -1))
         b = _view_row_as_element(max_channels.values.reshape(intensities.shape[0], -1))
 
-        # TODO ambrosejcarr: the object makes working with the data awkward
-        # we could store a map between targets and ints as an `attr`, and use that to convert
-        # for the public API.
-        targets = np.empty(intensities.shape[0], dtype=object)
-        targets.fill("None")
+        targets = np.full(intensities.shape[0], fill_value=np.nan, dtype=object)
 
-        for i in np.arange(a.shape[0]):
+        # decode the intensities
+        for i in np.arange(codes.shape[0]):
             targets[np.where(a[i] == b)[0]] = codes[Features.TARGET][i]
-        target_index = pd.Index(targets.astype('U'))
 
-        intensities[Features.TARGET] = (
-            Features.AXIS, target_index)
+        # a code passes filters if it decodes successfully
+        passes_filters = np.ones_like(targets, dtype=np.bool)
+        passes_filters[pd.isnull(targets)] = False
+
+        intensities[Features.TARGET] = (Features.AXIS, targets.astype('U'))
+        intensities[Features.DISTANCE] = (Features.AXIS, distance)
+        intensities[Features.PASSES_FILTERS] = (Features.AXIS, passes_filters)
```

Comment:
right now, we filter (hopefully to rename to threshold, see comment above) decoded features on area and intensity. we should ALSO filter on distance/quality. Perhaps submit an Issue for this?